### ISSUE TITLE
Allow to selectively skip Chucker interception

### ISSUE THREAD

**Issue by MiSikora** [ChuckerTeam/chucker#266]
## :warning: Is your feature request related to a problem? Please describe

Sometimes I might not want to involve Chucker in the interception process on some endpoints. For example I know that request or response will be too large so Chucker won't handle it anyway. Or there might be some bug preventing me from using Chucker on this endpoint but I don't want to downgrade it as there are some other features I like in the new version.

## :bulb: Describe the solution you'd like

Add a header to the request that would be detected by the `ChuckerInterceptor` and inform it to not process the request. `ChuckerInterceptor` would then strip this header and pass it to down the chain without processing it.

It might be something like this.

```kotlin
fun Service {
  @GET("/")
  @Headers(Chucker.SKIP_INTERCEPTOR)
  fun networkRequest(): Unit
}
```

## :raising_hand: Do you want to develop this feature yourself?

- [x] Yes
- [ ] No

**Comment by cortinico** [comment]
Reopening this issue as we had a discussion from PR #279, that causes its subsequent revert in #300 

tl;dr: We really like the feature but we should find a way to implement it without bleeding implementation details inside the `library-no-op` variant of Chucker.

Here I'm trying to summarise how the feature should behave from my point of view.

### On the `library` (debug) variant

Ideally here Chucker should just skip the processing of any request that contains the special header. Given the assumption that we're on a Debug environment, I will also not remove the header. If `ChuckerInterceptor` finds it, it will just pass over the request without doing any processing.

### On the `library-no-op` (release) variant

Ideally the `ChuckerInterceptor` here should do nothing at all. This is sort of part of the API contract that we have with our users. I'm afraid that adding this feature could create a precedent for adding other feature that requires request/response manipulation on the `-no-op` variant. Having problems on the `library-no-op` variant means delivering bug in production for app that are using Chucker as a debug tool, and this is something that we really don't want.

### Stripping out the Header

So the current approach with a custom header, poses the problem of **stripping the header** from the HTTP request. The alternatives I see so far are probably a combination of those points:

1. Write a message on the README file about this feature where we point out that header stripping is not automatic and should be implemented by the user.
2. Provide a snippet with a sample interceptor that they can use to strip the headers away (see below).
3. Create another module or create another repo that will host a `ChuckerHeaderStripperInterceptor` (name TBD) that will just perform the stripping.
4. Advice to apply the custom header only on `Debug` environment. This can be extremely inconvenient as developers that are most likely using Retrofit or other REST abstraction libraries, don't have an easy way to apply headers only on Debug.

Feel free to suggest more if you have any other idea.

Some final thoughts on every approach:

#### 1. README message

We will anyway add a section on the README for this feature. So a disclaimer will be added anyway. Probably having only the disclaimer is not enough but will definitely create inform the user.

#### 2. Sample Interceptor

Ideally point 1 can be improved with a snippet like this one:

```kotlin
class HeaderStripperInterceptor : Interceptor {
    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request().newBuilder()
                .removeHeader("SPECIAL HERE HERE")
                .build()
        return chain.proceed(request)
    }
}
```

We will also have to suggest also that this extra interceptor must be applied **after** Chucker otherwise the extra header will be useless.

#### 3. Adding another module/repo

This could be a nice addition but comes with the cost of having another module to maintain. The users will have to apply the extra dependency in a way like this:

```gradle
dependencies {
  debugImplementation "com.github.ChuckerTeam.Chucker:library:<version>"
  releaseImplementation "com.github.ChuckerTeam.Chucker:library-no-op:<version>"
  implementation "com.github.ChuckerTeam.Chucker:header-stripper:<version>"
}
```

With this, developers have to intentionally opt-in to have the header stripping also in production (and this is supported by the usage of the plain `implementation` configuration).

#### 4. Advising to add header only on Debug.

This can potentially be added as a follow-up for point 1. As I already said, it's probably not feasible for the majority of developer that are most likely using Retrofit or other libraries on top of OkHTTP.

**Comment by MiSikora** [comment]
One consideration might be to create also an extensions like this in the external library.

```kotlin
fun Interceptor.stripping(headerName: String): Interceptor
```

It would make it easier for Kotlin users to use this feature. Function could be mark with `@JvmSynthetic` to hide it from Java users. I think that it would be awkward to have such a static utility function.

**Comment by psh** [comment]
Suppose in the `no-op` variant, I have a class
```kotlin
class ChuckerCallFactory(private val wrapped: Call.Factory) : Call.Factory {
    override fun newCall(request: Request): Call = wrapped.newCall(request)

    fun addFilter(filter: (Request) -> Boolean) = Unit
}
```

that in the debug variant actually has a definition
```kotlin
class ChuckerCallFactory(private val wrapped: Call.Factory) : Call.Factory {
    private val skippedEndpoints: MutableList<(Request) -> Boolean> = mutableListOf()

    override fun newCall(request: Request): Call {
        val skip = skippedEndpoints.any { it(request) }
        return wrapped.newCall(
            if (skip) {
                request.newBuilder().addHeader("X-Chucker-Skip-Interception", "true").build()
            } else {
                request
            }
        )
    }

    @Suppress("unused")
    fun addFilter(filter: (Request) -> Boolean) {
        skippedEndpoints.add(filter)
    }
}
```
Then in our code somewhere we can configure our call factory with a predicate that says when to add the "skip" header to a request.  In this experiment, let's add to all `GET` requests - your lambda can filter on other request parameters if you want.
```kotlin
private val chuckerCallFactory = ChuckerCallFactory(httpClient).apply {
    // I don't care about GET request, please don't log them ;)
    addFilter { it.method().equals("get", true) }
}
```

And this can be configured in my `Retrofit` really easily:  The `Retrofit.Builder` code shows that the call to "client()" is actually an alias for "callFactory()", so we can insert our own -
```diff
private val api: HttpBinApi by lazy {
    Retrofit.Builder()
        .baseUrl(BASE_URL)
-        .client(httpClient)
+        .callFactory(chuckerCallFactory)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(HttpBinApi::class.java)
```
Final piece of the puzzle is the interceptor that strips out the header and skips processing if needed.
```kotlin
override fun intercept(chain: Interceptor.Chain): Response {
    val originalRequest = chain.request()
    val skip = originalRequest.header("X-Chucker-Skip-Interception") != null
    val request = originalRequest.newBuilder().removeHeader("X-Chucker-Skip-Interception").build()

    if (skip) {
        return chain.proceed(request)
    } else {
        // normal interception logic
    }
}
```
End result: 
* in release builds, there are no extra added headers, and the overhead is a method that acts as a direct pass-through (my _gut_ says that R8 will see that and inline it).  
* In the debug build we have the capability to specify any variety of logic for _which_ endpoints should be ignored by Chucker.
* If someone doesn't want to use the feature, and just log everything as normal, there's no change to the `Retrofit` configuration & no stray headers.  Backwardly compatible with all existing codebases.

**Comment by psh** [comment]
But, why not skip the header, and just configure the filters on the interceptor instead?
```kotlin
private val skippedEndpoints: MutableList<(Request) -> Boolean> = mutableListOf()

override fun intercept(chain: Interceptor.Chain): Response {
    val request = chain.request()

    if (skippedEndpoints.any { it(request) }) {
        return chain.proceed(request)
    } else {
        // normal interception logic
    }
}
```

**Comment by MiSikora** [comment]
Yes, I like this idea. Didn't think of using `CallFactory`.

**Comment by BukT0p** [comment]
Chucker is a tool, it has own db and UI, why not to implement filtering as part of the tool? Like you can have list of ignored items, you can manage that list, and it might be a good idea to make some mask and actions on urls to add to ignore list.

**Comment by vbuberen** [comment]
> Chucker is a tool, it has own db and UI, why not to implement filtering as part of the tool? Like you can have list of ignored items, you can manage that list, and it might be a good idea to make some mask and actions on urls to add to ignore list.

We appreciate contributions, so you are more than welcome to propose your changes in a PR.
